### PR TITLE
Add an hard reference boolean field to references. This is set by Ent…

### PR DESCRIPTION
**Pull request by galganif**
…ityBase if the schema field has the @hardRef annotation, and persisted in the database in the entity_refs table. It will be later used to bulk delete those entities.

### CODE REVIEW

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 1]
```diff
@@ -1933,13 +1941,6 @@ class DatabaseImpl(
...
         )
-        private val TABLES_VERSION_2 = TABLES_VERSION_1
-        private val TABLES_VERSION_3 = TABLES_VERSION_2
-        private val TABLES_VERSION_4 = TABLES_VERSION_3
-        private val TABLES_VERSION_5 = TABLES_VERSION_4
-
-        /* internal */
-        val TABLES = TABLES_VERSION_5
 
         private val CREATE_VERSION_3 =
```
Comment (shans) [inline java/arcs/android/storage/database/DatabaseImpl.kt:1945]: should we be updating the name of this to CREATE_VERSION_6?

**Review by shans — APPROVED** [review]

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 1 above)
Comment (galganif) [reply java/arcs/android/storage/database/DatabaseImpl.kt:1945]: Maybe I need to create a new create version, adding @jasonwyatt to take a look at the db migration

File: java/arcs/core/entity/Reference.kt [hunk 2]
```diff
@@ -32,6 +32,11 @@ class Reference<T : Entity>(
     val expirationTimestamp
         get() = storageReference.expirationTimestamp
 
+    val isHardReference
+        get() = storageReference.isHardReference
+
+    fun setHardReference() { storageReference.isHardReference = true }
```
Comment (jasonwyatt) [inline java/arcs/core/entity/Reference.kt:38]: Nit: don't one-line method bodies like this.. The only time they should be one-lined is if they're able to be expressed using function-expression notation.
```suggestion
    fun setHardReference() { 
      storageReference.isHardReference = true 
    }
```

File: java/arcs/core/storage/Reference.kt [hunk 3]
```diff
@@ -33,7 +33,8 @@ data class Reference(
     /** Reference creation time (in milliseconds). */
     private var _creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
     /** Reference expiration time (in milliseconds). */
-    private var _expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    private var _expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+    var isHardReference: Boolean = false
```
Comment (jasonwyatt) [inline java/arcs/core/storage/Reference.kt:37]: Some KDoc here to explain what a hard reference is would be helpful.

File: java/arcs/core/entity/Reference.kt [hunk 4]
```diff
@@ -32,6 +32,11 @@ class Reference<T : Entity>(
     val expirationTimestamp
         get() = storageReference.expirationTimestamp
 
+    val isHardReference
```
Comment (jasonwyatt) [inline java/arcs/core/entity/Reference.kt:35]: KDoc to explain what a hard reference is would be helpful.

File: javatests/arcs/android/storage/database/DatabaseImplTest.kt [hunk 5]
```diff
@@ -664,6 +664,41 @@ class DatabaseImplTest {
         assertThat(entityOut).isEqualTo(parentEntity)
     }
 
+    @Test
```
Comment (jasonwyatt) [inline javatests/arcs/android/storage/database/DatabaseImplTest.kt:667]: It'd be good to add a test to ensure the migration works as expected as well.

File: javatests/arcs/android/entity/HardReferenceTest.kt [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    @After
+    fun tearDown() {
+        WriteBackForTesting.clear()
+        schedulerProvider.cancelAll()
+    }
+
+    @Test
+    fun hardReferenceWorkEndToEnd() = runBlocking<Unit> {
+        val referencedEntityHandle = handleManager.createCollectionHandle(
+            referencedEntitiesKey, entitySpec = TestReferencesParticle_Entity_Referenced
```
Comment (jasonwyatt) [inline javatests/arcs/android/entity/HardReferenceTest.kt:89]:
```suggestion
            referencedEntitiesKey, 
            entitySpec = TestReferencesParticle_Entity_Referenced
```

File: javatests/arcs/android/entity/HardReferenceTest.kt [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        )
+        val child = TestReferencesParticle_Entity_Referenced(5)
+        referencedEntityHandle.dispatchStore(child)
+        val childRef = referencedEntityHandle.dispatchCreateReference(child)
+
+        val entity = TestReferencesParticle_Entity(referenced = childRef)
+        assertThat(entity.referenced!!.isHardReference).isTrue()
+
+        val writeHandle = handleManager.createCollectionHandle(
+            collectionKey, entitySpec = TestReferencesParticle_Entity
```
Comment (jasonwyatt) [inline javatests/arcs/android/entity/HardReferenceTest.kt:99]:
```suggestion
            collectionKey, 
            entitySpec = TestReferencesParticle_Entity
```

File: javatests/arcs/android/entity/HardReferenceTest.kt [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        )
+        writeHandle.dispatchStore(entity)
+
+        val readHandle = handleManager.createCollectionHandle(
+            collectionKey, entitySpec = TestReferencesParticle_Entity
```
Comment (jasonwyatt) [inline javatests/arcs/android/entity/HardReferenceTest.kt:104]:
```suggestion
            collectionKey, 
            entitySpec = TestReferencesParticle_Entity
```

File: javatests/arcs/android/entity/HardReferenceTest.kt [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun <T : Entity> EntityHandleManager.createCollectionHandle(
+        key: StorageKey,
+        entitySpec: EntitySpec<T>
+    ) = createHandle(
+        HandleSpec(
+            "name", HandleMode.ReadWrite, CollectionType(EntityType(entitySpec.SCHEMA)), entitySpec
+        ), key, Ttl.Infinite()
```
Comment (jasonwyatt) [inline javatests/arcs/android/entity/HardReferenceTest.kt:121]:
```suggestion
        HandleSpec(
            "name", 
            HandleMode.ReadWrite, 
            CollectionType(EntityType(entitySpec.SCHEMA)), 
            entitySpec
        ), 
        key, 
        Ttl.Infinite()
```

File: javatests/arcs/android/entity/HardReferenceTest.kt [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            "set-referencedEntities", TestReferencesParticle_Entity_Referenced.SCHEMA.hash
+        )
+    )
+    private val collectionKey = ReferenceModeStorageKey(
+        backingKey = DatabaseStorageKey.Persistent(
+            "entities", TestReferencesParticle_Entity.SCHEMA.hash
+        ), storageKey = DatabaseStorageKey.Persistent(
+            "set-ent", TestReferencesParticle_Entity.SCHEMA.hash
+        )
+    )
```
Comment (jasonwyatt) [inline javatests/arcs/android/entity/HardReferenceTest.kt:57]:
```suggestion
    private val collectionKey = 
        ReferenceModeStorageKey(
            backingKey = DatabaseStorageKey.Persistent(
                "entities", 
                TestReferencesParticle_Entity.SCHEMA.hash
            ), 
            storageKey = DatabaseStorageKey.Persistent(
                "set-ent", 
                TestReferencesParticle_Entity.SCHEMA.hash
            )
        )
```

File: javatests/arcs/android/entity/HardReferenceTest.kt [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+@RunWith(AndroidJUnit4::class)
+class HardReferenceTest {
+    private val schedulerProvider = SimpleSchedulerProvider(Dispatchers.Default)
+    private val referencedEntitiesKey = ReferenceModeStorageKey(
+        backingKey = DatabaseStorageKey.Persistent(
+            "referencedEntities", TestReferencesParticle_Entity_Referenced.SCHEMA.hash
+        ), storageKey = DatabaseStorageKey.Persistent(
+            "set-referencedEntities", TestReferencesParticle_Entity_Referenced.SCHEMA.hash
+        )
+    )
```
Comment (jasonwyatt) [inline javatests/arcs/android/entity/HardReferenceTest.kt:50]:
```suggestion
    private val referencedEntitiesKey = 
        ReferenceModeStorageKey(
            backingKey = DatabaseStorageKey.Persistent(
                "referencedEntities", 
                TestReferencesParticle_Entity_Referenced.SCHEMA.hash
            ), 
            storageKey = DatabaseStorageKey.Persistent(
              "set-referencedEntities", 
              TestReferencesParticle_Entity_Referenced.SCHEMA.hash
            )
        )
```

**Review by jasonwyatt — APPROVED** [review]
Looks good. Mostly just nits.

It would be nice to have a test to validate the migration works as expected, however.

File: javatests/arcs/android/storage/database/DatabaseImplTest.kt (hunk 5 above)
Comment (galganif) [reply javatests/arcs/android/storage/database/DatabaseImplTest.kt:667]: Added some in DatabaseUpgradeTest. Let me know if you were thinking something different.

**Pull request merged**